pulsar-websocket/src/main/java/org/apache/pulsar/websocket/AbstractWebSocketHandler.java

```diff
@@ -55,44 +56,56 @@ public AbstractWebSocketHandler(WebSocketService service, HttpServletRequest req
         request.getParameterMap().forEach((key, values) -> {
             queryParams.put(key, values[0]);
         });
-    }
 
-    @Override
-    public void onWebSocketConnect(Session session) {
-        super.onWebSocketConnect(session);
-        log.info("[{}] New WebSocket session on topic {}", session.getRemoteAddress(), topic);
+        checkAuth(response);
+    }
 
+    private void checkAuth(ServletUpgradeResponse response) {
         String authRole = "<none>";
         if (service.isAuthenticationEnabled()) {
             try {
                 authRole = service.getAuthenticationService().authenticateHttpRequest(request);
-                log.info("[{}] Authenticated WebSocket client {} on topic {}", session.getRemoteAddress(), authRole,
+                log.info("[{}] Authenticated WebSocket client {} on topic {}", request.getRemoteAddr(), authRole,
                         topic);
 
             } catch (AuthenticationException e) {
-                log.warn("[{}] Failed to authenticated WebSocket client {} on topic {}: {}",
-                        session.getRemoteAddress(), authRole, topic, e.getMessage());
-                close(WebSocketError.AuthenticationError);
+                log.warn("[{}] Failed to authenticated WebSocket client {} on topic {}: {}", request.getRemoteAddr(),
+                        authRole, topic, e.getMessage());
+                try {
+                    response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Failed to authenticate");
+                } catch (IOException e1) {
+                    log.warn("Failed to send error: {}", e1.getMessage());
+                }
                 return;
             }
         }
 
         if (service.isAuthorizationEnabled()) {
             try {
                 if (!isAuthorized(authRole)) {
-                    log.warn("[{}] WebSocket Client [{}] is not authorized on topic {}", session.getRemoteAddress(), authRole,
-                            topic);
-                    close(WebSocketError.NotAuthorizedError);
+                    log.warn("[{}] WebSocket Client [{}] is not authorized on topic {}", request.getRemoteAddr(),
+                            authRole, topic);
+                    response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Not authorized");
```

Comment:
Is SC_UNAUTHORIZED (401) appropriate here?
I think SC_FORBIDDEN (403) is more appropriate.

Reply:
Makes sense, fixed